### PR TITLE
[TxMetricsExpoter] Migrate to contractkit

### LINKED ISSUE #474: Migrate Transaction Metrics Explorer to the new contractkit
### Expected Behavior

Use ContractKit

### Current Behavior

Use WalletKit

### PR DESCRIPTION
**Pull request by mcortesi**
### Description

**Commits are split into different features to make review easier**

Migrates tx-metricx-exporter to contractKit, and adds ability to parse Blocks and Logs to contractKit

### Tested

- `yarn test` for utils
- launch with port forward to integration, and then faucet.

### Other changes

Several:

 - New features to @celo/utils
 - New deployment for tx-metrics-exporter
 - Improve Dockerfile
 - [contractKit] Add isListening & isSyncing
 - [contractKit] Comment MultiSig. Not being deployed

### Related issues

- Fixes #474 

### Backwards compatibility

Yes

### CODE REVIEW

File: packages/contractkit/src/base.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ export enum CeloContract {
   GasPriceMinimum = 'GasPriceMinimum',
   GoldToken = 'GoldToken',
   Governance = 'Governance',
-  MultiSig = 'MultiSig',
+  // MultiSig = 'MultiSig',
```
Comment (nambrot) [inline packages/contractkit/src/base.ts]: Can we just remove?

File: packages/contractkit/src/explorer/base.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+export interface ContractDetails {
+  name: string
+  address: Address
+  jsonInterface: ABIDefinition[]
+}
+
+export async function obtainKitContractDetails(kit: ContractKit) {
+  const coreContractDetails: ContractDetails[] = []
+  for (const celoContract of AllContracts) {
+    const contract = await kit._web3Contracts.getContract(celoContract)
```
Comment (nambrot) [inline packages/contractkit/src/explorer/base.ts:15]: Can we do a Promise.all with a `map` here to do this in parallel?

File: packages/contractkit/src/explorer/block-explorer.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+export async function newBlockExplorer(kit: ContractKit) {
+  return new BlockExplorer(kit, await obtainKitContractDetails(kit))
+}
+
+export class BlockExplorer {
+  private addressMapping: Map<Address, ContractMapping> = new Map()
+
+  constructor(private kit: ContractKit, readonly contractDetails: ContractDetails[]) {
+    for (const cd of contractDetails) {
```
Comment (nambrot) [inline packages/contractkit/src/explorer/block-explorer.ts]: As you know I'm a fan of filter/mapping here :)

File: packages/contractkit/src/explorer/block-explorer.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      this.addressMapping.set(cd.address, {
+        details: cd,
+        fnMapping,
+      })
+    }
+  }
+
+  async fetchBlockByHash(blockHash: string): Promise<Block> {
+    // TODO fix typing: eth.getBlock support hashes and numbers
+    const block = await this.kit.web3.eth.getBlock(blockHash as any, true)
```
Comment (nambrot) [inline packages/contractkit/src/explorer/block-explorer.ts]: you can just return `this.kit.web3.eth.getBlock(blockHash as any, true)` straight?

File: packages/contractkit/src/explorer/block-explorer.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    return block
+  }
+  async fetchBlock(blockNumber: number): Promise<Block> {
+    const block = await this.kit.web3.eth.getBlock(blockNumber, true)
+    return block
```
Comment (nambrot) [inline packages/contractkit/src/explorer/block-explorer.ts]: Same as above

File: packages/contractkit/src/explorer/block-explorer.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  async fetchBlockRange(from: number, to: number): Promise<Block[]> {
+    const results: Block[] = []
+    for (let i = from; i < to; i++) {
+      results.push(await this.fetchBlock(i))
```
Comment (nambrot) [inline packages/contractkit/src/explorer/block-explorer.ts:62]: We'll probably want to do this in parallel with `Promise.all?`

File: packages/contractkit/src/explorer/log-explorer.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+      const fnMapping: Map<string, ABIDefinition> = new Map()
+      for (const abiDef of cd.jsonInterface as ABIDefinition[]) {
+        if (abiDef.type === 'event') {
+          fnMapping.set(abiDef.signature, abiDef)
+        }
+      }
+
+      this.addressMapping.set(cd.address, {
+        details: cd,
+        logMapping: fnMapping,
```
Comment (nambrot) [inline packages/contractkit/src/explorer/log-explorer.ts:61]: Should `fnMapping` be called `logMapping`?

File: packages/contractkit/src/web3-contract-cache.ts [hunk 8]
```diff
@@ -7,7 +7,7 @@ import { newGasPriceMinimum } from './generated/GasPriceMinimum'
 import { newGoldToken } from './generated/GoldToken'
 import { newGovernance } from './generated/Governance'
 import { newLockedGold } from './generated/LockedGold'
-import { newMultiSig } from './generated/MultiSig'
+// import { newMultiSig } from './generated/MultiSig'
```
Comment (nambrot) [inline packages/contractkit/src/web3-contract-cache.ts]: Can we remove this?

File: packages/transaction-metrics-exporter/src/blockchain.ts [hunk 9]
```diff
@@ -1,215 +1,193 @@
...
-  if (parsedTransaction === null) {
-    return
-  }
+  const provider = kit.web3.currentProvider as WebsocketProvider
 
-  parsedTransactionCounter.inc({
-    contract: parsedTransaction.contractName,
-    function: parsedTransaction.functionName,
-  })
+  // TODO(mcortesi): verify if we need this (I guess we don't)
```
Comment (nambrot) [inline packages/transaction-metrics-exporter/src/blockchain.ts]: Remove

File: packages/transaction-metrics-exporter/src/blockchain.ts [hunk 10]
```diff
@@ -1,215 +1,193 @@
...
+
+          logEvent('RECEIVED_PARSED_LOG', event)
+        }
+      }
     }
   }
+}
 
-  pollConnectionStatus()
+/*
```
Comment (nambrot) [inline packages/transaction-metrics-exporter/src/blockchain.ts]: Remove

File: packages/utils/src/async.ts [hunk 11]
```diff
@@ -16,11 +21,33 @@ export const retryAsync = async (
...
 }
+
+/**
+ * Map an async function over a list xs with a given concurrency level
+ *
+ * @param concurrency number of `mapFn` concurrent executions
+ * @param xs list of value
+ * @param mapFn mapping function
+ */
+export async function concurrentMap<A, B>(
```
Comment (nambrot) [inline packages/utils/src/async.ts:40]: This seems unused?

File: packages/utils/src/async.ts [hunk 12]
```diff
@@ -1,5 +1,10 @@
 const TAG = 'utils/src/async'
 
+/** Wait for a number of milliseconds */
+export function wait(ms: number): Promise<void> {
```
Comment (nambrot) [inline packages/utils/src/async.ts]: We technically have a npm package for it `sleep-promise`, and we keep redefining it everywhere. Maybe this is a good place but definitely would like to stream line. Also we have traditionally called this function `sleep` which might be good to do here as well

**Review by nambrot — COMMENTED** [review]
Minor comments

File: packages/utils/src/async.ts (hunk 11 above)
Comment (mcortesi) [reply packages/utils/src/async.ts:40]: used now

File: packages/utils/src/async.ts (hunk 12 above)
Comment (mcortesi) [reply packages/utils/src/async.ts]: renamed to sleep.

I wouldn't recommend adding a package for something like this.

File: packages/protocol/migrations/16_elect_validators.ts [hunk 13]
```diff
@@ -9,6 +8,7 @@ import {
   sendTransactionWithPrivateKey,
 } from '@celo/protocol/lib/web3-utils'
 import { config } from '@celo/protocol/migrationsConfig'
+import { blsPrivateKeyToProcessedPrivateKey } from '@celo/utils/lib/bls'
```
Comment (nambrot) [inline packages/protocol/migrations/16_elect_validators.ts:11]: This seems unrelated to this PR

**Review by nambrot — APPROVED** [review]
Awesome!

File: packages/protocol/migrations/16_elect_validators.ts (hunk 13 above)
Comment (mcortesi) [reply packages/protocol/migrations/16_elect_validators.ts:11]: it is.... had to do it, to be able to change the Dockerfile. a bit of a rabbit hole
